python/pyspark/worker.py

```diff
@@ -110,9 +116,20 @@ def wrapped(key_series, value_series):
                 "Number of columns of the returned pandas.DataFrame "
                 "doesn't match specified schema. "
                 "Expected: {} Actual: {}".format(len(return_type), len(result.columns)))
-        arrow_return_types = (to_arrow_type(field.dataType) for field in return_type)
-        return [(result[result.columns[i]], arrow_type)
-                for i, arrow_type in enumerate(arrow_return_types)]
+
+        if not assign_cols_by_pos:
+            try:
+                # Assign result columns by schema name
+                return [(result[field.name], to_arrow_type(field.dataType))
+                        for field in return_type]
+            except KeyError:
```

Comment:
I think we want to be a little more careful here, for example, an `KeyError` in to_arrow_type could lead to unexpected behavior. 

How about sth like this:
```
if any(isinstance(name, basestring) for name in result.columns):
    return [(result[field.name], to_arrow_type(field.dataType)) for field in return_type]
else:
    return [(result.iloc[:,i], to_arrow_type(field.dataType)) for i, field in enumerate(return_type)]
```

Reply:
This seems ok to me since it's basically the same, but I don't think we need to worry about `to_arrow_type` throwing a `KeyError`.  Is there any particular reason you suggested handling position like this?

```
[(result.iloc[:,i], to_arrow_type(field.dataType)) for i, field in enumerate(return_type)]
```

To me it seems better to look up by column labels, how it is currently

```
[(result[result.columns[i]], to_arrow_type(field.dataType))
                for i, field in enumerate(return_type)]
```

Reply:
I think `result.iloc[:,i]` and `result[result.columns[i]]` are the same, you don't have change it if you prefer `result.columns[i]`

I agree `to_arrow_type` doesn't throw `KeyError`,  but in general I feel it's more robust not to assume the implementation detail of `to_arrow_type`. I think the code is more concise and readable with if/else too (comparing to except KeyError)